File: array/192.cpp
#include <bits/stdc++.h>

using namespace std;

bool isAdjacent(string a, string b) {
    if (a.size() != b.size()) return false;

    int cont = 0;
    for (int i = 0; i < a.size(); ++i)
        if (a[i] != b[i]) ++cont;

    return (cont > 1) ? false : true;
}

void BFS(string start, string target, unordered_map<string, int> &visited, int &cont,
         vector<string> &dictionary) {
    
    if (start == target) {
        return;
    }

    queue<pair<string, int>> q;
    q.push({start, 0});

    while (!q.empty()) {
        auto v = q.front();
        q.pop();

        if(v.first == target) {
            cont = min(cont, v.second);
            return;
        }

        visited[v.first] = 1;

        for (const auto &elem : dictionary) {
            if (visited[elem] == 0 && isAdjacent(v.first, elem)) {
                q.push({elem, v.second + 1});
            }
        }
    }
}

int main() {
    vector<string> dictionary = {"POON", "PLEE", "SAME", "POIE", "PLEA", "PLIE", "POIN"};
    string start = "TOON";
    string target = "PLEA";
    unordered_map<string, int> um;
    int cont = INT_MAX;

    for (int i = 0; i < dictionary.size(); ++i) um[dictionary[i]] = 0;

    BFS(start, target, um, cont, dictionary);

    cout << cont + 1<<"\n";

    dictionary.clear();
    dictionary = {"ABCD", "EBAD", "EBCD", "XYZA"};
    start = "ABCV";
    target = "EBAD";
    um.clear();
    cont = INT_MAX;

    for (int i = 0; i < dictionary.size(); ++i) um[dictionary[i]] = 0;

    BFS(start, target, um, cont, dictionary);

    cout << cont + 1<<"\n";

    return 0;
}

File: array/192.py
from typing import List
from collections import deque
import sys


def is_adjacent(a: str, b: str):
    cont = 0

    if len(a) != len(b):
        return False

    for s, t in zip(a, b):
        if s != t:
            cont += 1

    if cont > 1:
        return False

    return True


def bfs(start: str, target: str, dictionary: List, visited: dict, cont: List):
    if start == target:
        return

    q = deque([[start, 0]])

    while len(q) > 0:
        v = q.popleft()
        visited[v[0]] = 1

        if v[0] == target:
            cont[0] = min(cont[0], v[1])

        for word in dictionary:
            if visited[word] == 0 and is_adjacent(v[0], word):
                q.append([word, v[1] + 1])


dictionary = ["POON", "PLEE", "SAME", "POIE", "PLEA", "PLIE", "POIN"]
start = "TOON"
target = "PLEA"
visited = {item: 0 for item in dictionary}
cont = [sys.maxsize]
bfs(start, target, dictionary, visited, cont)
print(cont[0] + 1)


dictionary = ["ABCD", "EBAD", "EBCD", "XYZA"]
start = "ABCV"
target = "EBAD"
visited = {item: 0 for item in dictionary}
cont = [sys.maxsize]
bfs(start, target, dictionary, visited, cont)
print(cont[0] + 1)
